File: ruby/lib/security.rb
require 'services/common/v1/common_pb'
require 'okapi/security/v1/security_pb'
require 'okapi'
require 'okapi/hashing'
require 'okapi/hashing/v1/hashing_pb'
require 'okapi/oberon'

module Trinsic
  # Interface
  module ISecurityProvider
    def get_auth_header(account_profile, message)
      raise "Not implemented"
    end
  end

  class OberonSecurityProvider
    include ISecurityProvider

    def base64_binary_encode(message)
      Base64.urlsafe_encode64(Google::Protobuf::encode(message))
    end

    def get_auth_header(account_profile, message)
      if account_profile.protection.enabled
        raise "The token must be unprotected before use"
      end

      request_hash = Google::Protobuf::encode(message)
      request_hash = Okapi::Hashing.blake3_hash(Okapi::Hashing::V1::Blake3HashRequest.new(data: request_hash)).digest unless request_hash.length == 0 # skip hashing if empty
      nonce = Trinsic::Common_V1::Nonce.new(timestamp: (Time.now.to_f*1000).to_int, request_hash: request_hash)
      request = Okapi::Security::V1::CreateOberonProofRequest.new(token: account_profile.auth_token,
                                                                  data: account_profile.auth_data,
                                                                  nonce: Google::Protobuf::encode(nonce))
      proof = Okapi::Oberon.create_proof(request)

      # Convert to base-64
      "Oberon ver=#{1},proof=#{Base64.urlsafe_encode64(proof.proof)},data=#{Base64.urlsafe_encode64(account_profile.auth_data)},nonce=#{base64_binary_encode(nonce)}"
    end
  end
end

File: go/services/security.go
package services

import (
	"encoding/base64"
	"fmt"
	"github.com/trinsic-id/okapi/go/okapi/hashing/v1/hashing"
	"github.com/trinsic-id/okapi/go/okapi/security/v1/security"

	"time"

	"github.com/trinsic-id/okapi/go/okapi"
	"github.com/trinsic-id/sdk/go/proto/services/account/v1/account"
	"github.com/trinsic-id/sdk/go/proto/services/common/v1/common"

	"google.golang.org/protobuf/proto"
)

// SecurityProvider defines the required functionality to provide authentication to the api
type SecurityProvider interface {
	GetAuthHeader(profile *account.AccountProfile, message proto.Message) (string, error)
}

// OberonSecurityProvider implements the SecurityProvider interface and provides oberon token functionality
type OberonSecurityProvider struct {
}

// GetAuthHeader returns an authentication header with a correctly formatted oberon token
func (o OberonSecurityProvider) GetAuthHeader(profile *account.AccountProfile, message proto.Message) (string, error) {
	if profile != nil && profile.Protection.Enabled {
		return "", fmt.Errorf("the token must be unprotected before use")
	}

	requestBytes, err := proto.Marshal(message)
	if err != nil {
		return "", err
	}

	hashResult, err := okapi.Hashing().Blake3Hash(&hashing.Blake3HashRequest{Data: requestBytes})
	if err != nil {
		return "", err
	}
	requestHash := hashResult.Digest
	nonce := &common.Nonce{
		Timestamp:   time.Now().UnixMilli(),
		RequestHash: requestHash[:],
	}
	nonceBytes, err := proto.Marshal(nonce)
	if err != nil {
		return "", err
	}

	var authData, authToken []byte
	if profile != nil {
		authData = profile.AuthData
		authToken = profile.AuthToken
	}

	proof, err := okapi.Oberon().CreateProof(&security.CreateOberonProofRequest{
		Data:  authData,
		Token: authToken,
		Nonce: nonceBytes,
	})
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("Oberon ver=%d,proof=%s,data=%s,nonce=%s", 1, base64.URLEncoding.EncodeToString(proof.Proof), base64.URLEncoding.EncodeToString(profile.AuthData), base64.URLEncoding.EncodeToString(nonceBytes)), nil
}
